text/0000-placement-by-return.md

```diff
@@ -0,0 +1,926 @@
+    Struct2 { member: q }
+}
+```
+
+In other words, Rust does not currently guarantee Named Return Value Optimization.
+
+In the rustonomicon, it should mention that the `write_return_with` intrinsic can be used to build a function that's equivalent to `bad()`:
+
+```rust
+use std::mem::{write_return_with, ReturnSlot};
+use std::alloc::Layout;
+fn not_bad() -> [i32] {
+    unsafe {
+        write_return_with(Layout::new::<[i32; 1_000_000]>(), |arr: *mut u8| {
+            let arr: &mut [i32] = slice::from_raw_parts_mut(arr, 1_000_000);
+            for i in 0..1_000_000 {
+                arr[i] = i;
+            }
+            arr
+        })
+    }
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Did you say I can return unsized types?
+
+A function that directly returns an unsized type should be compiled into two functions, essentially as a special kind of generator:
+
+```rust
+// sugar
+fn my_function() -> str {
+    *"hello world"
+}
+fn function_that_calls_my_function() -> str {
+    println!("Hi there!");
+    my_function()
+}
+
+// desugar my_function
+// this is written in order to aid understanding, not because these are good APIs
+use std::alloc::Layout;
+struct __MyFunction__Internal {
+  local_0: &'static str,
+}
+impl __MyFunction__Internal {
+  /* "unsize coroutines" are different from normal generators because their behavior is much more
+    restricted. They yield exactly one Layout, and then they are finished by writing to a pointer. */
+  unsafe fn start(state: *mut __MyFunction__Internal) -> Layout {
+    /* As usual for generators, locals (including anonymous ones) get desugared into member
+      variables of an anonymous struct. In this case, the local variable is just a string literal. */
+    state.local_0 = "hello world";
+    Layout::for_value("hello world")
+  }
+  unsafe fn finish(state: *mut __MyFunction__Internal, slot: *mut u8) -> &mut str {
+    ptr::copy(state.local_0, slot.as_mut_ptr(), mem::size_of_val(state.local_0));
+    /* Notice how I also have to return a properly-built fat pointer?
+       This isn't very important for string slices, but it's mandatory for trait objects, because I need to
+       supply a vtable. */
+    str::from_raw_parts_mut(slot, state.local_0.len())
+  }
+}
+
+// desugar function_that_calls_my_function
+use std::alloc::Layout;
+struct __FunctionThatCallsMyFunction__Internal {
+  delegate: __MyFunction__Internal,
+}
+impl __FunctionThatCallsMyFunction__Internal {
+  unsafe fn start(state: *mut __FunctionThatCallsMyFunction__Internal ) -> Layout {
+    /* first, run everything leading up to the return */
+    println!("Hi, there!");
+    /* then, desugar the return (in this case, by delegating */
+    __MyFunction__Internal::start(&mut state.delegate)
+  }
+  unsafe fn finish(state: *mut __FunctionThatCallsMyFunction__Internal, slot: *mut u8) -> &mut str {
+    __MyFunction__Internal::start(&mut state.delegate, slot)
```

Comment:
```suggestion
    __MyFunction__Internal::finish(&mut state.delegate, slot)
```

Reply:
Oh, right. I remember being puzzled by that line when I first read it. Completely forgot about it.